### PR TITLE
Some changes on validators

### PR DESCRIPTION
**Pull request by diegonvs**
## shapeOf:

Instead returns a simple message containing a constraint in a shape validation, now returns an validation error that property in shape returns.

If shapeOf catches a error, `validator()` function can be called returning a error that others validators inside shapeOf have returned.

Example: 

```
Error: Invalid state passed to 'icon.a'. Expected type 'object', but received type 'number'. Passed to 'ClayButton’.
```

## arrayOf:
	
Instead return a simple constraint now returns a error showing that a index in array validator returns. 

I’ve modified `validateArrayItems()` function that emits arrayOf error message, this function now returns a validator call in the same way of shapeOf and i’ve appended a string for show the error message on the error’s index. 

Example:

```
Error: Invalid state passed to 'values'. Validator for values[2] says: "Error: Invalid state passed to 'values'. Expected type 'string', but received type 'number'. Passed to 'ClayDropdown'. " Passed to 'ClayDropdown'. 
```

### 'Warning':

I've removed warning string fragments to compose a message more clean instead of:

```
Warning: Error: Warning: Invalid state passed to 'label'. Expected type 'string', but received type 'number'. Passed to 'ClayBadge'. 
```
After changes will be like:
```
Error: Invalid state passed to 'label'. Expected type 'string', but received type 'number'. Passed to 'ClayBadge'. 
```

### Some fixes:

On solving these problems, i’ve noticed that shapeOf validation doesn’t verify `shape`(argument of function), causing that developer only receive a feedback of bad using this function on instantiate a new instance of metal. I had solved these problem validating arguments before executing them like, but i think this is not appropriated handling for devs.

I’m searching for enhance this handling. What do you think guys?

### CODE REVIEW

File: packages/metal-state/src/validators.js [hunk 1]
```diff
@@ -135,7 +144,7 @@ const validators = {
 				}
 				if ((required && !isDefAndNotNull(value[key])) ||
 					isInvalid(validator(value[key]))) {
-					return composeError(ERROR_SHAPE_OF, name, context);
+					return validator(value[key], name + '.' + key, context);
```
Comment (matuzalemsteles) [inline packages/metal-state/src/validators.js]: Hey @diegonvs, Here we could use the literal of ES6.
```javascript
`${name}.${$key}`
```

File: packages/metal-state/test/Config.js [hunk 2]
```diff
@@ -124,6 +124,34 @@ describe('Config', function() {
...
+				'propOne',
+				'propTwo',
+				'propThree'
+			]).value('propOne'),
+			four: Config.number().required()
+		};
+		var configShape = Config.arrayOf(Config.shapeOf(shape));
+		assert.ok(core.isObject(configShape));
+		assert.ok(core.isFunction(configShape.config.validator));
+		assert.ok(configShape.config.validator({
```
Comment (matuzalemsteles) [inline packages/metal-state/test/Config.js]: Why use `configShape.config.validator` not only needed to pass `configShape({...})`?

File: packages/metal-state/src/validators.js (hunk 1 above)
Comment (diegonvs) [reply packages/metal-state/src/validators.js]: done!

File: packages/metal-state/test/Config.js (hunk 2 above)
Comment (diegonvs) [reply packages/metal-state/test/Config.js]: I've based on other tests in config and discussing with @matuzalemsteles 
personally, i'll keep this.

**Comment by brunobasto** [comment]
Looks good! Thank you, @diegonvs!

The only thing I noticed is that the code is not formatted according to the project rules. Running "npm run format" changes a bunch of files (not only the ones you changed).

"npm run lint" is also failing. Can you put that on your backlog?

**Comment by diegonvs** [comment]
Okay @brunobasto! i'll check sf and verify lint task!

**Pull request closed without merging**